📖 add proposal for opt-in autoscaling from zero (#4283)
kubernetes-sigs/cluster-api: docs/proposals/20210310-opt-in-autoscaling-from-zero.md

---
title: Opt-in Autoscaling from Zero
authors:
  - "@elmiko"
reviewers:
  - "@"
creation-date: 2021-03-10
last-updated: 2021-03-10
status: implementable
---

# Opt-in Autoscaling from Zero

## Table of Contents

- [Title](#title)
  - [Table of Contents](#table-of-contents)
  - [Glossary](#glossary)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals/Future Work](#non-goalsfuture-work)
  - [Proposal](#proposal)
    - [User Stories](#user-stories)
      - [Story 1](#story-1)
      - [Story 2](#story-2)
    - [Security Model](#security-model)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan-optional)
  - [Implementation History](#implementation-history)

## Glossary

* **Node Group** This term has special meaning within the cluster autoscaler, it refers to collections
  of nodes, and related physical hardware, that are organized within the autoscaler for scaling operations.
  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled

>   differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond

Reviewer: IIRC, currently there is no functional difference between a Cluster Autoscaler at a management level and a workload cluster level apart from some basic configuration (please correct me if this is incorrect).
Now this introduces a feature which is available only in the capi world, and non-capi clusters will not have this functionality directly. Was there a reason why this was not done as a cluster-autoscaler feature and more as a capi feature?

Author: _some_ of the other providers support scale from zero (i think AWS and GCP), but this is a feature which is available in the upstream for any provider who wishes to implement. it is part of the [node group interface](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L172) and is optional to implement.

> Was there a reason why this was not done as a cluster-autoscaler feature and more as a capi feature?

yes, essentially it must be done by each provider. the advantage for CAPI is that through this provider you will get scale from zero on every platform that CAPI supports. adding this functionality to the other providers in the autoscaler would be nice, but it is a little beyond scope for this effort.

  directly to MachineSets and MachineDeployments that are marked for autoscaling.

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
scaling on Cluster API deployed clusters for MachineSets and MachineDeployments. One feature
that is missing from this integration is the ability to scale down to, and up from, a MachineSet
or MachineDeployment with zero replicas.

> This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define

Reviewer 2: I woud suggest to use infrastructure providers instead of cloud providers, given that AFAIK this proposal targets all the CAPI infrastructure providers, not only the ones running on a real cloud infrastructure.
```suggestion
This proposal defines an opt-in mechanism whereby Cluster API users and infrastructure providers can define
```
Same down in the document

Author: good suggestion, i will change those references

- Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.

### Non-Goals/Future Work

- Create an API contract that cloud providers must follow.
- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
- Support for MachinePools, either with the cluster autoscaler or using cloud provider native implementations (eg AWS AutoScalingGroups).
- Create an autoscaling custom resource for Cluster API.
- Add MachinePool support to the cluster autoscaler.

## Proposal

To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
this information to create a prediction about how many nodes should be created when scaling. In
most situations this information can be directly read from the nodes that are running within a
node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
autoscaler needs to acquire this information from the cloud provider.

> 

Reviewer 3: @fabriziopandini @elmiko 

Maybe an idea to address the concern mentioned by Fabrizio here: https://github.com/kubernetes-sigs/cluster-api/pull/4283#r593444976

What about only using the annotations on both the InfrastructureMachineTemplates and the MachineDeployment, ....

Possible advantages:
* CAPI types do not have to be changed for this, so it would be basically a feature build on top of CAPI APIs by autoscaler via annotations
* Only autoscaler implementation has to be adjusted
* Because of this it would work with all infrastructure providers out of the box (automatically reconciling the data on InfrastructureMachineTemplates could still be implemented, but it would also work without any changes to any InfraStructureProvider)

I think that would be a good option to address the concern that we spread autoscaler concerns into CAPI APIs.

Author: this is certainly an approach we could take, i had avoided going down this path further as i had been advised that annotations are a very weak form of API.

this is how we do scale from zero in openshift today, we use an annotation on the machineset to inform the autoscaler about the sizing requirements. i would be happy to contribute this solution, i think it could be even easier than suggested above. we could allow annotations to exist on either a machinedeployment or machineset, no need for the infrastructure template (as it would require another call to the api server from inside the autoscaler).

Reviewer 3: I think if we have those fields/annotations only MachineDeployments/MachineSets or also on InfrastructureMachineTemplates is something which can be decided independently from if we use annotations or fields.

So what I was proposing was to have annotations on MachineDeployments/MachineSets and InfrastructureMachineTemplates (if still desired).

I agree with that the annotations are a weaker form of API (mostly because they are not as type safe as regular fields), but to me it feels like it would be an ideal way to layer the autoscaler API on top of the CAPI types. Let's see what others think about it.

Author: yeah, i'm totally fine with this approach and it makes it very easy since we already have this implemented for openshift. slightly tangential, but i would like to see us create a new resource that can be used for cluster autoscaling, so maybe in the end it will be better for us to use a lightweight solution now. 

thanks for the feedback @sbueringer , i have added this as a topic for the meeting today. maybe we can come to a consensus.

without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
creation of nodes that will not consume resources until they are required by applications on my cluster.

### Implementation Details/Notes/Constraints

Cloud providers, or users, can add an optional field to their machine template infrastructure objects to indicate
the resources provided by the machine. Internally, this field will be represented by a Go `map` type
utilizing named constants for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values
(similar to how resource limits and requests are handled for pods).

**Example implementation in Docker provider**
```
// these constants will be carried in Cluster API, but are repeated here for clarity
const (
    AutoscalerResourceCPU ResourceName = "cpu"
    AutoscalerResourceMemory ResourceName = "memory"
    AutoscalerResourceGPU ResourceName = "gpu"
)

// DockerMachineTemplateSpec defines the desired state of DockerMachineTemplate.
type DockerMachineTemplateSpec struct {

>     AutoscalerCapacity ResourceList `json:"autoscalerCapacity,omitempty"`

Reviewer 2: I know that it introduces some nesting, some struct nesting, but should this be `spec.autoscaler.machine.request`...

Author: i had thought about something like this, but wasn't sure how much nesting we wanted XD

i will sketch up a version that uses this structure so we can see how it looks.

Reviewer 3: This information is only used when scaling from zero right? I wonder if there's a way to make this more explicit in the field names that it's only used in that situation (and e.g. not when scaling further up)

Author: yeah, it's a tough problem from a naming standpoint. i had originally thought using the word "hint" might be helpful, but after looking through the kubernetes docs, i feel like "capacity" is probably closer to what we are asking for here.

    Template DockerMachineTemplateResource `json:"template"`
}
```
_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_

When used as a manifest, it would look like this:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: DockerMachineTemplate
metadata:
  name: workload-md-0
  namespace: default
spec:

>   autoscalerCapacity:

Reviewer: I have a fundamental question. If the same MachineTemplate is used throughout my cluster but I support many compute configurations (different cpu, memory etc), do I need to create a set of machine templates, one for each compute configuration?
Alternatively, could we make this section into a list of compute configurations so that it is evident that it is the same template with different configurations?

Author: each MachineTemplate will describe one configuration (eg 4 CPU, 64 GB ram), so you would need to create multiple templates to describe different configurations, but you could share 1 template among MachineDeployments that describe the same type of machine.

>  Alternatively, could we make this section into a list of compute configurations so that it is evident that it is the same template with different configurations?

this would make less sense given that the spec is only describing one configuration. imo, in the future we might have a more autoscaling specific resource which could contain a list of these things, but for this PR it is just a single machine.

>     memory: 500mb

Reviewer 4: can this field be also called memoryMb and user only inputs a number just like the override annotation on MachineSet?

Author: an interesting thought, i think we definitely could do that but currently the implementation is using a parsed resource quantity here. so, the user could put bytes or use an SI prefix. restricting it to `Mb` might make things clearer, especially with the annotation.

i would like to look at the code a little more to see what implications it might have to change this.

Reviewer 2: +1 to be consistent with Kubernetes resource asks (e.g Pod request/limits or resource quotas)

Reviewer 3: Oh yes, let's please keep this consistent with requests/limits

Author: ok, so just to close the loop here. i will keep these names as they currently are(but might change the outer structure) and they will adhere to taking [resource.Quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/#Quantity) values, similar to Pod requests.

    cpu: "1"
    gpu: "1"
  template:
    spec: {}
```

> If a user wishes to override the resources hints as provided from the machine template, they may

Reviewer 4: just curious what does `hints` refer to?

Author: that is the previous name i had been using, i need to update that. thanks!

do so by adding the following annotations to their MachineSets or MachineDeployments:

```

> kind: MachineSet

Reviewer 2: TBH I have some concern about the entire idea of overrides.
- Afaik templates de facto are limited to a cluster/to a machineDeployment due to how we define ownerRef and how cascading delete works, so the infrastructure machine is already specific to a set of machines. 
- If we want to allow overrides on one of the CAPI resources, should those go on the MachineDeployment instead, given that MachineSet are kind of an internal detail e they comes and go automatically, thus forcing the user to re-apply override every time a new MS gets created
- Having two places where the same information can be set could be kind of confusing

What about unblocking the scale for zero problem and then iterate on UX as soon as we have more concrete feedback from the field?

Author: >     * Afaik templates de facto are limited to a cluster/to a machineDeployment due to how we define ownerRef and how cascading delete works, so the infrastructure machine is already specific to a set of machines.

perhaps i misunderstood something then, is it possible to have a single infrastructure machine template that is linked in 2 separate machineDeployments?

the other thing about this is that the autoscaler resource requests in the infrastructure machine are optional, so the overrides would provide an extremely lightweight way to add scale from zero to a single group of machines without having to change the infrastructure.

>     * If we want to allow overrides on one of the CAPI resources, should those go on the MachineDeployment instead, given that MachineSet are kind of an internal detail e they comes and go automatically, thus forcing the user to re-apply override every time a new MS gets created

my thought here is that allowing it on both resources makes it easy to use for multiple styles of user.

>     * Having two places where the same information can be set could be kind of confusing

yeah, i think that's fair. we would document with examples, but there is no escaping that this would be the same information in multiple places (if utilized).

> What about unblocking the scale for zero problem and then iterate on UX as soon as we have more concrete feedback from the field?

i could be convinced to drop this ;)

but, on the ux front, these annotations are how we use scale from zero currently with machinesets on openshift. i think it would be cool if we contribute that part of feature as well. i appreciate the feedback Fabrizio!

Reviewer 3: I think it's possible to add it to several MachineDeployments but not to multiple clusters. The InfrastructureMachineTemplate has a ownerReferences to the cluster, i.e. when the cluster is deleted the InfrastructureMachineTemplate is also deleted.

last-updated: 2021-03-10
status: implementable
---

# Opt-in Autoscaling from Zero

## Table of Contents

- [Title](#title)
  - [Table of Contents](#table-of-contents)
  - [Glossary](#glossary)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals/Future Work](#non-goalsfuture-work)
  - [Proposal](#proposal)
    - [User Stories](#user-stories)
      - [Story 1](#story-1)
      - [Story 2](#story-2)
    - [Security Model](#security-model)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan-optional)
  - [Implementation History](#implementation-history)

## Glossary

* **Node Group** This term has special meaning within the cluster autoscaler, it refers to collections
  of nodes, and related physical hardware, that are organized within the autoscaler for scaling operations.
  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled
  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond

>   directly to MachineSets and MachineDeployments that are marked for autoscaling.

Author: the autoscaler does not currently know about machine pools. we will have to make some decisions about how to do that integration, but i left them out because currently they are not evaluated directly from the autoscaler side.

unless i'm misunderstanding ;)

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
scaling on Cluster API deployed clusters. One feature that is missing from this integration is
the ability to scale down to, and up from, a MachineSet or MachineDeployment with zero replicas.

This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define

> the specific resource requirements for each Infrastructure Machine Temalate they create. These

Reviewer 5: typo `Temalate`

requirements are then utilized by the cluster autoscaler to make decisions about the number

> of nodes to add during scale up operations.

Reviewer 5: To be clear this information is only used by the autoscaler when scaling up from zero, as in such scenario there's no node resource to get the info.

## Motivation

Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
implemented for many of the integrated cloud providers. It is a popular feature that has been
requested for Cluster API on multiple occasions. This feature empowers users to reduce their
operational resource needs, and likewise reduce their operating costs.

Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
implementations, this feature might not make sense in all scenarios. To accomodate the wide
range of deployment options in Cluster API, the scale to zero feature will be optional for
users and cloud providers.

### Goals

- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.

> - Provide a mechanism for users to override the defaults for any given MachineSet or MachineDeployment.

Reviewer 5: which defaults is this referring to?

Author: i guess "defaults" is ambiguous here. i mean to say that there will be a mechanism for users to override the values provided in the infrastructure machine template.

- Create an optional API contract that defines the minimum needed information for scaling from and to zero replicas.

### Non-Goals/Future Work

- Create an API contract that cloud providers must follow.
- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
- Create additional CRDs or controllers

## Proposal

To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
this information to create a prediction about how many nodes should be created when scaling. In

> most sitatuations this information can be directly read from the nodes that are running within a

Reviewer 5: typo `sitatuations`

node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the

> autoscaler needs to acquire this information from the cloud provider.

Author: the problem is more on the autoscaler side with how it does calculations about scaling up. doing the _1 to n_ scaling is easier because the autoscaler can evaluate the resources used by the nodes running to calculate how many nodes should be added.
when the autoscaler sees a node group with size 0, it needs some information to instruct about how to calculate the number of needed nodes to satisfy the unsheduled pods. for the other providers, these are just lookup tables based on the instance types in the node group. but for capi the linkage is not as straightforward.

> autoscaler needs to acquire this information from the cloud provider.

Reviewer 2: > it needs some information to instruct about how to calculate the number of needed nodes to satisfy the unsheduled pods

I don't have direct knowledge of the autoscaler internals, but what is kind of strange, is that because the autoscaler can't manage "lack of informations", we are spreading autoscaler concerns across other APIs.

If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

> autoscaler needs to acquire this information from the cloud provider.

Author: > I don't have direct knowledge of the autoscaler internals, but what is kind of strange, is that because the autoscaler can't manage "lack of informations", we are spreading autoscaler concerns across other APIs.

i think this is impacting us more directly than any other cloud provider for the autoscaler though. mainly due to our architecture. for example, the other cloud providers have implementations that live within the autoscaler repository. so for them, this information is compiled directly into the autosacler (eg AWS has a lookup table with this information).

> If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

it might be possible for us to cache this information inside our implementation at the autoscaler, but i would worry about it getting lost on a restart. i'll give this some thought.

> autoscaler needs to acquire this information from the cloud provider.

Reviewer 5: >If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

Autoscaler falling back to 1 would completely defeat the purpose of the feature since you only want to scale up if the upcoming nodes are a good fit for the workloads to be scheduled.
The contract for any provider to be autoscalable from zero is to provide this information therefore the need to expose it so it can be consumed.

> autoscaler needs to acquire this information from the cloud provider.

Reviewer 2: Fall back to 1 will serve only for moving the autoscaler out of a situation that it can manage properly (it is required to scale up, but there is not enough info to determine how many machine do we actually need).

My assumption, is that immediately after scaling to 1, we are back in a situation the autoscaler is already capable to manage properly, and thus the existing logic will kick in and add further nodes as required.
But might be am I wrong..
@elmiko to confirm/deny

> autoscaler needs to acquire this information from the cloud provider.

Author: > My assumption, is that immediately after scaling to 1, we are back in a situation the autoscaler is already capable to manage properly, and thus the existing logic will kick in and add further nodes as required.

this is correct. the autoscaler core essentially uses an interface to ask the provider-specific bits for a "node group template" (a template of what the instances in that node group look like). when a node group is at zero we will use the information from the resource structure, but if the node group has a size greater than zero, then the autoscaler will use a node group template based on the nodes that already exist.

> autoscaler needs to acquire this information from the cloud provider.

Reviewer 2: thanks for context.
the idea is to have something similar:

- if the replicas is 0, and it is required to scale up:
   - start scaling to 1 (given that we don't know how many machines are actually required due to lack of machines capacity info) 
   
then, as soon as the machine capacity are available, the autoscaler will determine if to add more replicas (as of today)

> autoscaler needs to acquire this information from the cloud provider.

Author: that is more or less how it works

> An optional field is proposed for users to add to their Infrastructure Machine Templates which will

Reviewer 6: If that field is similar in all infrastructure providers, is there something blocking us from having it directly in MachineSet spec? I understand that infrastructure template can be reused but won't having it in MS spec make it more obvious that it is scaled from zero? WDYT?

Author: in general, i tend to agree with you. it would be nice to have this information only on the MachineSet or MachineDeployment resources, but one problem here is that we want this information to be more closely related to the individual infrastructure providers since they will need to supply this data.

when we discussed this issue previously, one issue that arose was the idea that each cloud provider should be able to opt-in to the scale from zero behavior depending on how their cloud implementation is structured. by placing this information in the Machine[Set|Deployment] we are moving it further from the provider implementations and closer to the users.

the idea that we had settled on during previous discussions was that the individual providers could populate this field on their infrastructure templates (depending on the machine sizes available in that cloud), and then it would be reconciled into the Machine[Set|Deployment] status. in this manner we expose the information on the Machine[Set|Deployment] but also make it clear that it is fixed information at that level.

the hope here is that we will reduce the need for users to directly control these values and allow them to be part of the infrastructure providers implementation details.

carry information about the CPU, memory, and GPU requirements for the machines created from the template.
This field will then be copied to the status of any MachineSet or MachineDeployment with an infrastructure
reference to the template during reconciliation. The cluster autoscaler will then read this information
from the status field to perform its calculations.

A user may override the field in the associated infrastructure template by applying annotations to the
MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
information in favor of reading the information from the status.

By keeping this information with the MachineSet and MachineDeployment resource objects, we reduce the

> need for the cluster autoscaler to have a deep knowledge of Cluster API internals.

Author: > This is easier from an autoscaler point of view, but this could be potentially confusing from a user perspective, because the source of truth for those information is the InfrastructureMachine, while the override happens on a different resource.

i tend to agree, and i think if we decide as a group that the best solution here is for the autoscaler to read the infrastructure resource from the MachineSet|Deployment, that is a viable option as well. but, we need to ensure that we will keep the infrastructure ref that currently exists.

> Also, from a user perspective, what is the difference/the need for having annotation at two levels (MachineDeployment, MachineSet)? Are there precedence rules?

i've used the language of "MachineSet or MachineDeployment" because the autoscaler could be operating on either type, depending on how the user has configured the autoscaler. but in practice, the autoscaler won't be dealing with both types of objects for the same node group. the user would only need to annotate the object they have specified for autoscaling.

> need for the cluster autoscaler to have a deep knowledge of Cluster API internals.

Author: i've been playing around with an implementation that does this copying and i'm starting to think it might be more appropriate for the autoscaler to extract the infrastructure template object if it has a reference, and then use annotations on the resource in question (eg MachineSet, MachineDeployment, and MachinePool somday) for an override mechanism.

so, from the capi side, a provider would only need to supply these fields on their infrastructure template and then the rest would work as expected. but, there would be no replication of the resource hint fields from the infrastructure ref to the status of the affected resource.

### User Stories

- Detail the things that people will be able to do if this proposal is implemented.
- Include as much detail as possible so that people can understand the "how" of the system.

> - The goal here is to make this feel real for users without getting bogged down.

Author: yes, thanks!

#### Story 1

As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.

#### Story 2

As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
creation of nodes that will not consume resources until they are required by applications on my cluster.

### Implementation Details/Notes/Constraints

A new common API type will be created named `ResourceHints`, its definition will look like this:

```

> type ResourceHints struct {

Author: i had originally been calling this `AutoscalerResourceHints`, but had opted for the more generic term to leave possibility for future expansion. i would be fine to scope it more tightly.

> type ResourceHints struct {

Author: @fabriziopandini just curious, would want to see something like this?

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4                                                                                                                
kind: DockerMachineTemplate                                                                                                                                         
metadata:                                                                                                                                                           
  name: workload-md-0                                                                                                                                               
  namespace: default                                                                                                                                                
spec:                                                                                                                                                               
  autoscaler: 
    resources:                                                                                                                                                   
      memory: 500mb                                                                                                                                                   
      vCPU: "1"                                                                                                                                                       
      GPU: "1"                                                                                                                                                        
  template:                                                                                                                                                         
    spec: {}
```

> type ResourceHints struct {

Author: @fabriziopandini just curious what you think here ^^

    vCPU   *string `json:"vCPU,omitempty"`
    Memory *string `json:"memory,omitempty"`
    GPU    *string `json:"gpu,omitempty"`
}
```

This new API type can be used with a machine template to inform about the resources used
by individual machines in the group. For example, here is a `DockerMachineTemplate` with

> the integrated `ResourceHints`

Reviewer 2: Are we expecting the user to provide those info or the cloud provider to fill in those info given the machine type selected by the user?

Author: this depends on how the user is consuming CAPI. i would expect that whoever is creating the infra machine templates for the deployment would be responsible for adding this information. it would be nice if the samples created by the provider implementation owners would demonstrate the usage of these fields.

how do our users handle the infra machine template currently?

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: DockerMachineTemplate
metadata:
  name: workload-md-0
  namespace: default
spec:

>   resourceHints:

Reviewer 2: Wondering if this field name should somehow link explicitly to the autoscaler, in order to avoid potential confusion in the user that might assume those field could be used to override the machine type selection or as way to customise default machine sizes.

    memory: 500mb
    vCPU: "1"
    GPU: "1"
  template:
    spec: {}
```

During reconciliation of MachineSets and MachineDeployments, if the `resourceHints` field is found
on a related `infrastructureRef`, then the `resourceHints` will be added to the `status` field
for that resource. For example, a MachineDeployment status might look like this:

```
status:
  observedGeneration: 1
  phase: Running
  selector: cluster.x-k8s.io/cluster-name=workload,cluster.x-k8s.io/deployment-name=workload-md-0
  resourceHints:
    memory: 500mb
    vCPU: "1"
    GPU: "1"
```

If a user wishes to override the resources hints as provided from the machine template, they may
do so by adding the following annotations to their MachineSets or MachineDeployments:

```
kind: MachineSet
metadata:
  annotations:

>       machine.x-k8s.io/GPU: "0"

Reviewer 2: Also in this case, wondering if the annotation should refer to the autoscaler explicitly

Author: i think that makes sense, i'll think about some name ideas.

throught the MachineSets and MachineDeployments.

Creating clear documentation about the flow information, and the action of the cluster autoscaler
will be the first line of mitigating the confusion around this process. Additionally, adding an
example in the test Docker provider will help to clarify usage for cloud providers.

## Alternatives

An alternative approach to reconciling the resource hints from the machine template to MachineSets
and MachineDeployments would be for the cluster autoscaler to directly read the machine template
through the `infrastructureRef` that exists within those objects. This would require the autoscaler
to query for the machine templates, as well as the required permissions for those objects.

Another alternative would be for cloud providers to dynamically provide this information at run time.
This may be an approach to consider in the future, but it would still require a method to expose
the data at the MachineSet and MachineDeployment level. In effect this might look like an
automated method of achieving the same action as this proposal.

A much larger alternative would be to create a new custom resource that would act as an autoscaling

> abstraction. This new resource would be accessed by both the cluster autoscaler and the Cluster API

Reviewer 2: I'm generally very conservative about adding a new resource, but in this case I kind of like the idea of replicating the same pattern used for HorizontalPodAutoscaler vs Deployments&ReplicaSets.
This open up to several interesting extensions, like maintenance mode, min max replicas, create/delete policies etc.

Author: i tend to agree. i think a great future path (and one that i will callout in an addition to this doc) is to work towards creating this autoscaling resource that could be consumed by the autoscaler and owned by a separate controller to facilitate the reconcile logic.

controllers, as well as potentially another operator to own its lifecycle. This approach would
provide the cleanest separation between the components, and allow for future features in a contained
environment. The downside is that this approach requires the most engineering and design work to
accomplish.

## Upgrade Strategy

> As this field is optional, it should not negatively affect upgrades. That said, care should be taken

Reviewer 2: Background (silly) question:
Is there a risk of conflicts between a MachineDeployment rolling out new replicas and an autoscaler trying managing the same MachineDeployment, especially now that we are implementing new scale in/scale out algorithms?

Author: i'll have to think about this a little more. in general, the autoscaler will not try to do anything until it sees either unscheduled pods or under-utilized nodes, this might be an area where we would want to disable autoscaling on a MachineDeployment if it is re-deploying a set of machines.

to ensure that this field is copied during any object upgrade as its absence will create unexpected
behavior for end users.

## Additional Details

### Test Plan

The cluster autoscaler tests for Cluster API integration do not currently exist outside of the downstream
testing done by Red Hat on the OpenShift platform. There have talks over the last year to improve this
situation, but it is slow moving currently.

The end goal for testing is to contribute the scale from zero tests that currently exist for OpenShift
to the wider Kubernetes community. This will not be possible until the testing infrastructure around
the cluster autoscaler and Cluster API have resolved more.

## Implementation History

- [X] 06/10/2021: Proposed idea in an issue or [community meeting]

> - [ ] MM/DD/YYYY: Compile a Google Doc following the CAEP template (link here)

Reviewer 2: You can reference the previous proposal as a preliminary work, or remove entirely this line

about the nodes, the resource requests are utilized to predict the number of nodes needed. This
information is only used by the autoscaler when it is scaling from zero.

## Motivation

Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
implemented for many of the integrated cloud providers. It is a popular feature that has been
requested for Cluster API on multiple occasions. This feature empowers users to reduce their
operational resource needs, and likewise reduce their operating costs.

Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
implementations, this feature might not make sense in all scenarios. To accomodate the wide
range of deployment options in Cluster API, the scale to zero feature will be optional for
users and cloud providers.

### Goals

- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
- Create an optional API contract in the Infrastructure Machine Template that allows cloud providers to specify
  instance resource requirements that will be utilized by the cluster autoscaler.

> - Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.

Reviewer 3: Probably a dumb question. Why do we need this override mechanism? Can't the user just change the value in the infrastructure machine template?

Author: my thinking here is that by allowing overrides on individual MachineSets/MachineDeployments, we give users a fine grain control to override just a specific node group without needing to modify the template for all node groups that use it.

Reviewer 3: Thx for the explanation, makes now sense to me.

### Non-Goals/Future Work

- Create an API contract that cloud providers must follow.
- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
- Support for MachinePools, either with the cluster autoscaler or using cloud provider native implementations (eg AWS AutoScalingGroups).
- Create an autoscaling custom resource for Cluster API.

## Proposal

To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
this information to create a prediction about how many nodes should be created when scaling. In
most situations this information can be directly read from the nodes that are running within a

> node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the

Reviewer 3: ```suggestion
node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
```

    AutoscalerResourceMemory corev1.ResourceName = "memory"
)

// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
type AutoscalerMachineStatus struct {
    Capacity corev1.ResourceList `json:"capacity,omitempty"`
}

// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
type DockerMachineTemplateStatus struct {
    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
}

// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
type DockerMachineTemplate struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
    Status DockerMachineTemplateStatus `json:"status,omitempty"`
}
```
_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_

When used as a manifest, it would look like this:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: DockerMachineTemplate
metadata:
  name: workload-md-0
  namespace: default
spec:
  template:
    spec: {}
status:

>   machine:

Reviewer 3: @elmiko Did you also want to drop machine?  (Looked like this to me in the other conversation)

(I think to drop machine as the resource is a MachineTemplate, so we don't have to have a `machine` layer here

Author: that makes sense, i had kept machine on purpose but i agree with your reasoning

    capacity:
      memory: 500mb
      cpu: "1"
      nvidia.com/gpu: "1"
```

#### MachineSet and MachineDeployment Annotations

In cases where a user needs to provide specific resource information for a
MachineSet or MachineDeployment, or in cases where an infrastructure provider
has not yet added the Infrastructure Machine Template status changes, they
may use annotations to provide the information. The annotation values match the
API that is used in the Infrastructure Machine Template, e.g. the memory and cpu
annotations allow a `resource.Quantity` value and the two gpu annotations allow
for the count and type of GPUs per instance.

If a user wishes to specify the resource capacity through annotations, they
may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
that are participating in autoscaling:

```
kind: <MachineSet or MachineDeployment>
metadata:
  annotations:
      capacity.cluster.x-k8s.io/gpu-count: "1"
      capacity.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"
      capacity.cluster.x-k8s.io/memory: "500mb"

>       capacity.cluster.x-k8s.io/cpu: "1"

Reviewer 7: Following up from yesterday's community meeting, maybe these annotations could be defined in Autoscaler directly and supported there?

Reviewer 8: In what sense? I think this was always going to be on the CA side in some way.

Reviewer 8: Oh, as in not have them live in the `cluster.x-k8s.io` namespace?

Author: i was going to remove the annotations from this enhancement since it sounded like we wanted to not have another point for the api interaction. do we want to keep them?

Reviewer 7: Up to you, my worry was mostly around exposing these annotations in Cluster API, and under the cluster.x-k8s.io domain

Author: ok, let me do a little research on the autoscaler side. if it makes sense then we could keep them with something like `capacity.autoscaler.x-k8s.io` ?

Reviewer 7: Works for me if those variables are only exposed on the autoscaler side

Author: i'm not fully following, what do you mean by "only exposed on the autoscaler side" ?

Reviewer 3: I think that the constants are added to the autoscaler codebase and not to CAPI.

- Create an optional API contract in the Infrastructure Machine Template that allows infrastructure providers to specify
  instance resource requirements that will be utilized by the cluster autoscaler.
- Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.

### Non-Goals/Future Work

- Create an API contract that infrastructure providers must follow.
- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
- Support for MachinePools, either with the cluster autoscaler or using infrastructure provider native implementations (eg AWS AutoScalingGroups).
- Create an autoscaling custom resource for Cluster API.

## Proposal

To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
is the CPU and memory resources for nodes within the target node group that will be scaled. The autoscaler
uses this information to create a prediction about how many nodes should be created when scaling. In
most situations this information can be directly read from the nodes that are running within a

> node group. But, during a scale from zero situation (ie when a node group has zero replicas) the

Author: > When exactly does the autoscaler access this information? Only if there are exactly 0 replicas?

that is correct. 

> E.g. if a user changes the annotation, currently existing nodes should not be affected. Only if we are at 0 replicas the annotation will be read.

yes, in the situations where there are 1 or more replicas in the node group, it will use those nodes to predict the sizing of new nodes.

controllers and will require more coordination between developers and users. The second method
requires less direct intervention from infrastructure providers and puts more resposibility on users, for
this additional responsibility the users gain immediate access to the feature. These methods are
mutually exclusive, and the annotations will take preference when specified.

#### Infrastructure Machine Template Status Updates

Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
This field will contain the CPU, memory, and GPU resources associated with the machine described by
the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
limits and requests are handled for pods).

It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
Each infrastructure provider will be responsible for determining the best implementation for adding the
status field based on the information available on their platform.

> 

Reviewer 3: Just out of curiosity, and probably no concern for the proposal:

I assume those fields would then be documented as optional in the machine contract here? 
https://cluster-api.sigs.k8s.io/developer/architecture/controllers/machine.html

Do we plan to add the types and annotation constants in one of the CAPI API packages (or add a new one) or somewhere in the autoscaler code base?

Author: > Just out of curiosity, and probably no concern for the proposal:
> 
> I assume those fields would then be documented as optional in the machine contract here?
> https://cluster-api.sigs.k8s.io/developer/architecture/controllers/machine.html
> 

yes, i think it would make sense to have this in the cluster-api book as well as the autoscaler readme page for capi.

> Do we plan to add the types and annotation constants in one of the CAPI API packages (or add a new one) or somewhere in the autoscaler code base?

they will be in the autoscaler package. i'm open to having them in the cluster-api code as well, but that would require more synchronization, although it might be easier for implementors.

Reviewer 3: Thx for the additional context. Agree, could be something that is helpful for infra providers if they standardize based on those structs (annotations are If I got it right only used by users or the autoscaler)

**Example implementation in Docker provider**
```
// these constants will be carried in Cluster API, but are repeated here for clarity
const (
    AutoscalerResourceCPU ResourceName = "cpu"

>     AutoscalerResourceMemory ResourceName = "memory"

Reviewer 3: Could `storage` also be relevant in the context of the autoscaler? (just saw that there's `corev1.ResourceName` for it)

Author: storage is not relevant in this context. when nodes are created then the autoscaler has a different method for detecting attached storage and storage requests.

    AutoscalerResourceGPU ResourceName = "gpu"
)

// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
type AutoscalerMachineStatus struct {

>     Request ResourceList `json:"request,omitempty"`

Reviewer 3: I wonder if `request` is the right word. Maybe something like `capacity` or `resources` fits better.

I know where it comes from, but I think in case of a Pod it has not quite the same meaning as here and maybe doesn't fit.

`autoscaler.machine.resources` or `autoscaler.machine.capacity` looks like a good fit to me

Author: i agree, i like `capacity` but i changed it to `request` to be more inline with @fabriziopandini 's suggestion. i'm happy to change it back.

Reviewer 3: @fabriziopandini wdyt? All options are fine for me.

    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
    Status DockerMachineTemplateStatus `json:"status,omitempty"`
}
```
_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_

When used as a manifest, it would look like this:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: DockerMachineTemplate
metadata:
  name: workload-md-0
  namespace: default
spec:
  template:
    spec: {}
status:
  autoscaler:
    machine:
      request:
        memory: 500mb
        cpu: "1"
        gpu: "1"
```

#### MachineSet and MachineDeployment Annotations

In cases where a user needs to provide specific resource information for a
MachineSet or MachineDeployment, or in cases where an infrastructure provider
has not yet added the Infrastructure Machine Template status changes, they
may use annotations to provide the information.

If a user wishes to specify the resource requests through annotations, they
may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
that are participating in autoscaling:

```
kind: <MachineSet or MachineDeployment>
metadata:
  annotations:

>       autoscaler-machine-request.x-k8s.io/gpu: "1"

Reviewer 3: Are those annotations already specified in the autoscaler? If not, should we use the Cluster API namespace?

Author: they are not, i was trying to suggest a name that would make it clear.  what if i change to:
```
cluster.x-k8s.io/autoscaler-machine-request-gpu
```

and if we like `capacity` better for the status block, i would change the annotation name accordingly.

Reviewer 3: WDYT about: `autoscaler.cluster.x-k8s.io/machine-request-gpu`?
But I'm fine with whatever you can get consensus on :)

I just think we probably shouldn't use `x-k8s.io` without further namespacing.

Author: makes sense to me, i will wait until we decide the other naming question and then fix this as well. thanks!

>       autoscaler.cluster.x-k8s.io/machine-capacity-memoryMb: "500"

Reviewer 3: ```suggestion
      autoscaler.cluster.x-k8s.io/machine-capacity-memory: "500mb"
```
nit: I wonder if we should have the unit in the value and not in the annotation name (would be consistent with the field / the pod request/limit fields)

Reviewer 3: (last nit from my side, I promise :))

Author: my feeling here is that if we drop the `Mb` from the name then the value should be whatever a `resource.Quantity` could handle. adding it to the annotation key makes it explicit what is expected here.

Reviewer 3: Yup, just think its inconsistent with the corresponding field. But feel free to keep it as is.

Author: i don't have a strong feeling about keeping the `Mb` in the annotation, but i want to confer with a colleague to see if they had another reason for suggesting that name. if not, i'm happy to change it and just say we will allow `resource.Quantity` in that field (i suppose the cpu field would follow the same pattern then).

Reviewer 3: Oh, good point that cpu could also be a quantity. I'm not sure if cpu has to be an integer though. Are there cases where we don't have "whole" cpus?

Author: ok, had some conversations about these annotations and have update the memory one to be more consistent with the structured API.

This proposal describes opt-in mechanisms whereby Cluster API users and infrastructure providers can define
the specific resource requirements for each Infrastructure Machine Template they create. In situations
where there are zero nodes in the node group, and thus the autoscaler does not have information
about the nodes, the resource requests are utilized to predict the number of nodes needed. This
information is only used by the autoscaler when it is scaling from zero.

## Motivation

Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
implemented for many of the integrated infrastructure providers. It is a popular feature that has been
requested for Cluster API on multiple occasions. This feature empowers users to reduce their
operational resource needs, and likewise reduce their operating costs.

Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
implementations, this feature might not make sense in all scenarios. To accomodate the wide
range of deployment options in Cluster API, the scale to zero feature will be optional for
users and infrastructure providers.

### Goals

> - Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.

Reviewer 5: nit: Provide capability for Cluster API MachineSets and MachineDeployments to be auto scaled from and to zero replicas by the cluster autoscaler.

Scaling from/to zero expressing intent, is possible today.

most situations this information can be directly read from the nodes that are running within a
node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
autoscaler needs to acquire this information from the infrastructure provider.

An optional status field is proposed on the Infrastructure Machine Template which will be populated
by infrastructure providers to contain the CPU, memory, and GPU capacities for machines described by that
template. The cluster autoscaler will then utilize this information by reading the appropriate
infrastructure reference from the resource it is scaling (MachineSet or MachineDeployment).

A user may override the field in the associated infrastructure template by applying annotations to the
MachineSet or MachineDeployment in question. This annotation mechanism also provides users an opportunity
to utilize scaling from zero even in situations where the infrastructure provider has not given the information
in the Infrastructure Machine Template. In these cases the autoscaler will evaluate the annotation
information in favor of reading the information from the status.

### User Stories

#### Story 1

> As a Cluster API user, I would like to reduce my operating costs by scaling down my workload

Reviewer 5: who is this user/persona, a cluster operator?

Reviewer 5: same for the other stories

Reviewer 8: Suggested some replacements.

Author: thanks for the suggestions @randomvariable !

> As a Cluster API user, I would like to reduce my operating costs by scaling down my workload

Reviewer 8: ```suggestion
As a cloud administrator, I would like to reduce my organisation's operating costs by scaling down application workloads
```

Although I realise we should add some more personas to
https://github.com/kubernetes-sigs/cluster-api/blob/58905fb5140a7844c3e24aa70d255fb8f9d0c491/docs/book/src/user/personas.md

cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.

#### Story 2

> As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
> without the need for human intervention. As these nodes might be more expensive, I would also like to return them when

Reviewer 8: ```suggestion
As an application developer, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
```

not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
creation of nodes that will not consume resources until they are required by applications on my cluster.

#### Story 3

> As a Cluster API user, I would like to have access to the scale from zero feature but my infrastructure provider

Reviewer 8: ```suggestion
As a cluster operator, I would like to have access to the scale from zero feature but my infrastructure provider
```

this additional responsibility the users gain immediate access to the feature. These methods are
mutually exclusive, and the annotations will take preference when specified.

#### Infrastructure Machine Template Status Updates

Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
This field will contain the CPU, memory, and GPU resources associated with the machine described by
the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
limits and requests are handled for pods).

It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
Each infrastructure provider will be responsible for determining the best implementation for adding the
status field based on the information available on their platform.

**Example implementation in Docker provider**
```
// these constants will be carried in Cluster API, but are repeated here for clarity
const (

>     AutoscalerResourceCPU ResourceName = "cpu"

Reviewer 8: Let's change it to `corev1.ResourceName` as per https://github.com/kubernetes-sigs/cluster-api/pull/4283/files#r714797847

>     AutoscalerResourceCPU ResourceName = "cpu"

Author: thanks again, i wasn't sure the best place to grab these from. appreciate the guidance =)

    AutoscalerResourceMemory ResourceName = "memory"
    AutoscalerResourceGPU ResourceName = "gpu"
)

// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
type AutoscalerMachineStatus struct {
    Capacity ResourceList `json:"capacity,omitempty"`
}

// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
type AutoscalerMachineTemplateStatus struct {
    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
}

// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
type DockerMachineTemplateStatus struct {
    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
}

// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
type DockerMachineTemplate struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
    Status DockerMachineTemplateStatus `json:"status,omitempty"`
}
```

> _Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_

Reviewer 8: ```suggestion
_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
```
nit: we probably don't want to import k/k

Author: > nit: we probably don't want to import k/k

completely agree!

When used as a manifest, it would look like this:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: DockerMachineTemplate
metadata:
  name: workload-md-0
  namespace: default
spec:
  template:
    spec: {}
status:

>   autoscaler:

Reviewer 7: Do we need the autoscaler prefix here?

Reviewer 3: @vincepri Looking at: https://github.com/kubernetes-sigs/cluster-api/pull/4283/files#r716762327

Does it make sense to drop both the autoscaler and capacity prefix/layer?

Author: this structure is what we arrived at after some discussion. i had originally had these values with fewer layers wrapping them, but during the review we added the extra scoping to make it clear. if we want to go back on that i'm ok with, but we should get buy-in from @fabriziopandini as well.

Reviewer 3: I don't have strong opinions either way, so just ignore me :)

Reviewer 2: What about having simply status.capacity=

Author: i'm fine with that, i think that brings us back around to a simpler/flatter structure for this. if folks are ok with `status.capacity`, i will adjust the enhancement accordingly.

Reviewer 7: +1 to drop the autoscaler prefix, just to add more color about my comment here and in the annotation suggestion is that in the future we might want to be able to reuse this information or bring it elsewhere as well (machines for example, by inspecting nodes, etc)

Author: i'll change the references for the status block and annotations, thanks for the continued discussion.

    machine:
      capacity:
        memory: 500mb
        cpu: "1"
        gpu: "1"
```

> #### MachineSet and MachineDeployment Annotations

Reviewer 7: Should these annotations be set on MachineSet or MachineDeployment rather than the infrastructure template reference?

Author: yes, these are only for MachineSets and MachineDeployments, not infrastructure templates. i thought i had made it clear, but should i add more language to specify they are not for infrastructure templates?

Reviewer 7: I was wondering if it'd actually be more appropriate to have them on infrastructure template resources rather than MS/MD

Author: i think my preference would be to leave it on the MS/MD so that users could override with a finer granularity (eg same template for 2 different MS/MD). we talked about this a little in the past, but i don't think we discussed moving the annotations to the infrastructure templates.

In cases where a user needs to provide specific resource information for a
MachineSet or MachineDeployment, or in cases where an infrastructure provider
has not yet added the Infrastructure Machine Template status changes, they
may use annotations to provide the information. The annotation values match
API that is used in Infrastructure Machine Template, e.g. the memory and cpu
annotations allow a `resource.Quantity` value and the gpu annotation allows the
count of GPUs per instance.

If a user wishes to specify the resource capacity through annotations, they
may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
that are participating in autoscaling:

```
kind: <MachineSet or MachineDeployment>
metadata:
  annotations:

>       autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"

Reviewer 8: Referring back to, can we make this example work for namespaced resources in https://github.com/kubernetes-sigs/cluster-api/pull/4283#issuecomment-901280727 as it will start to  become important in cloud providers that support different GPU architectures.

Reviewer 8: Oh, ok. Is the GPU type going to be another annotation then?

Reviewer 8: Yes, that's right 
![image](https://user-images.githubusercontent.com/1441070/134410647-1a4ff0de-fa99-4ae7-9e96-9fb95339b45f.png)

Author: > Is the GPU type going to be another annotation then?

i don't think we would need it as the simulator is able to determine the GPU type when it does the node creation simulation. (i am still understanding this code better, but i've seen it work)

> Slightly off-topic info which might be relevant: I'm not sure about annotations, but you can only have one / in labels, so this could also apply to annotations.

i might have misunderstood @randomvariable , but i was expecting we would allow this type of usage
```yaml
metadata:
  annotations:
      autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "nvidia.com/gpu: 1"
```
but now that i type it out, it looks way too confusing.

Reviewer 8: >  the simulator is able to determine the GPU type when it does the node creation simulation

Isn't this simply the case because there's provider specific code that maps instance types to GPU types? The AWS code also doesn't seem to know about AMD GPUs so presumably that's not supported at the moment. There seems to be a bit more logic for GKE if I'm reading the code correctly.

That said, I'm still trying to get my head around the cluster autoscaler space, so I'm probably missing something.

It's probably unlikely you'd have two different GPU types on the same host, so could probably do:

`autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"`
`autoscaler.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"`

Which would work, but are we carrying baggage from the autoscaler's past here? The device plugin architecture is supposed to generic. In the on-premise space, we're starting to see nVidia suggest the appearance HPC workloads that use networking resources as well as GPUs, in particular Mellanox resources provided by Network Operator https://docs.mellanox.com/display/COKAN10/Network+Operator . In these cases we then start to care about RDMA & GPU resources, e.g. a pod with the declared requests:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: testpod1
  annotations:
    k8s.v1.cni.cncf.io/networks: sriov-net
spec:
  containers:
  - name: appcntr1
    image: <image>
    imagePullPolicy: IfNotPresent
    securityContext:
      capabilities:
          add: ["IPC_LOCK"]
     volumeMounts:
      - mountPath: /dev/hugepages
        name: hugepage
      resources:
        requests:
          memory: 1Gi
          hugepages-1Gi: 2Gi
    command: [ "/bin/bash", "-c", "--" ]
    args: [ "while true; do sleep 300000; done;" ]
    resources:
      requests:
        nvidia.com/rdma_host_dev: '1'
        nvidia.com/gpu: '1'
      limits:
        nvidia.com/rdma_host_dev: '1'
        nvidia.com/gpu: '1'
  volumes:
   - name: hugepage
     emptyDir:
       medium: HugePages
```
This all seems fine from what has been proposed for the inframachinetemplate, since the definition of capacity can be a map. But this doesn't seem to work well for the manual annotations.

Author: i think the questions you raise are very poignant, and sadly i'm not sure i can answer them to the required depth necessary.

what i can say is that we currently run the autoscaling from zero on openshift with the annotations the way they exist and i've seen it do the proper thing when there are workload pending which have requested `nvidia.com/gpu: 1`.

> That said, I'm still trying to get my head around the cluster autoscaler space, so I'm probably missing something.

you and me both, i'm sure ;)

i am still diving through the simulator code to better understand how this works at the low level, but so far the best i can explain is that it appears that inside the simulator code is where it determines the exact type of GPU and if the pending pods will fit. my understanding is that it uses pieces of the core scheduler code to do this work, and that there is also a stage where it _predicts_ what the nodes will look like. it is during this prediction part that it find the appropriate resources. that's the best description i can give at the moment.

> It's probably unlikely you'd have two different GPU types on the same host, so could probably do:
>
> `autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"`
> `autoscaler.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"`
>
> Which would work, but are we carrying baggage from the autoscaler's past here?

yeah, that could be possible. plus, i don't think adding that extra annotation would necessarily make any difference. my current observations are that this works with a simple numeric representation of the count of GPUs available. i'd prefer to keep this more simple for now and then add to it if we need to.

>  In the on-premise space, we're starting to see nVidia suggest the appearance HPC workloads that use networking resources as well as GPUs, in particular Mellanox resources provided by Network Operator https://docs.mellanox.com/display/COKAN10/Network+Operator . In these cases we then start to care about RDMA & GPU resources, e.g. a pod with the declared requests:

this is an excellent call out and something that we are seeing as well. i'm not sure if this is being handled just yet in the autoscaler, but i think our best option here is to keep consistent with what the autoscaler currently wants and help to make this better as we see the need. i don't have a ton of insight into the other accelerators just yet as i have been focused on the GPU stuff, but there are extension points within the autoscaler (eg custom node processors) which i imagine will help with this effort.

> This all seems fine from what has been proposed for the inframachinetemplate, since the definition of capacity can be a map. But this doesn't seem to work well for the manual annotations.

another great point, although my current understanding is that even with the expanded resource capacity status we are adding in this enhancement it will still boil down to a "number of GPUs present" at the API level during the initial node group pruning.

my preference is to stick with the annotations as they currently are, but i will add the `gpu-type` option if this is a blocker for you.

Author: @randomvariable i double checked the code and you are correct, we will need to have a gpu-type field as well. i'm going to update again.

      autoscaler.cluster.x-k8s.io/machine-capacity-memory: "500mb"

>       autoscaler.cluster.x-k8s.io/machine-capacity-cpu: "1"

Reviewer 7: ```suggestion
      capacity.cluster.x-k8s.io/cpu: "1"
```
Maybe?

Author: i'd be ok with this, @sbueringer and @fabriziopandini inspired the original change, i'd like to get their opinion as well.

Reviewer 3: I'm not sure what we had before the current one. It would be fine for me to shorten it.

Reviewer 4: I think we should keep the word `autoscaler` in the annotation as this annotation is on the MD/MS object, so without that word the purpose of these annotations might not become clear. So either what it is right now, or `autoscaler.cluster.x-k8s.io/cpu: "1"` if we want to shorten it

Reviewer 3: I think the reason for dropping autoscaler is actually to make the annotation also usable outside of the autoscaler use case (especially for the struct in status). Otherwise it would always be a field which is only there for the autoscaler.

Reviewer 4: oh ok I thought it was meant to be only for autoscaler. 
> especially for the struct in status

I didn't get this part?

Reviewer 3: I just meant the same applies to the `status.autoscaler.machine`. You would never use those fields for other use cases then autoscaler. 
Note: I'm just guessing about @vincepri's intentions :)

Author: currently this information is just for the autoscaler, but if i understand @sbueringer 's intention he is saying that in the future we might want to re-use these annotations for other things.

i think @vincepri 's suggestion is good for `capacity.cluster.x-k8s.io/<resource>`, we had discussed previously having the `machine` part in there as well to help indicate that is was per-machine. but if the community is happy with a short name, i think it makes the ux a little easier.

Author: > Note: I'm just guessing about @vincepri's intentions :). But for me it comes down to, either we want to make clear those fields are dedicated for the autoscaler or we open them up to other use cases. (I don't know what we want)

+1, i think the important information to convey is that these are resource capacity information about machines in the group. adding any mention of the "autoscaler" is helpful for understanding the usage, but is it something we want to bake into the API? (i can see the wisdom is dropping it, but we had requests previously to add it)

Reviewer 2: no strong opinions, but if we are removing autoscaler from status, let's remove it from here as well

Reviewer 3: > no strong opinions, but if we are removing autoscaler from status, let's remove it from here as well

+1. I also only think we should do it consistently

Author: agreed about keeping them consistent
